src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -493,56 +530,134 @@ private Float[][] PostProcess(Float[][] y, Float[] sigma, Float[] z, int d, int
         }
 
-        protected override ColumnType GetColumnTypeCore(int iinfo)
+        internal static void ValidatePcaInput(IHost host, string name, ColumnType type)
         {
-            Host.Check(0 <= iinfo & iinfo < Utils.Size(_types));
-            return _types[iinfo];
+            if (!type.IsVector)
+                throw host.Except($"Pca transform can only be applied to vector columns. Column ${name} is of type ${type}");
+
+            if (!(type.IsKnownSizeVector && type.VectorSize > 1))
+                throw host.Except($"Pca transform can only be applied to vector columns. Column ${name} is of size ${type.VectorSize}");
+
+            var itemType = type.ItemType;
+            if (itemType.RawKind != DataKind.R4)
+                throw host.Except($"Pca transform can only be applied to vector of float items. Column ${name} contains type ${itemType}");
         }
 
-        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
+        private sealed class Mapper : MapperBase
         {
-            Host.AssertValueOrNull(ch);
-            Host.AssertValue(input);
-            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
-            disposer = null;
-
-            var getSrc = GetSrcGetter<VBuffer<Float>>(input, iinfo);
-            var src = default(VBuffer<Float>);
-            var trInfo = _transformInfos[iinfo];
-            ValueGetter<VBuffer<Float>> del =
-                (ref VBuffer<Float> dst) =>
+            public sealed class ColumnSchemaInfo
+            {
+                private readonly string _input;
+                private readonly string _output;
+                private readonly string _weightColumn;
+                private readonly Schema _schema;
+
+                public ColumnSchemaInfo((string input, string output) columnPair, Schema schema, string weightColumn = null)
                 {
-                    getSrc(ref src);
-                    TransformFeatures(Host, ref src, ref dst, trInfo);
-                };
-            return del;
-        }
+                    _input = columnPair.input;
+                    _output = columnPair.output;
+                    _weightColumn = weightColumn;
+                    _schema = schema;
+                }
 
-        private static void TransformFeatures(IExceptionContext ectx, ref VBuffer<Float> src, ref VBuffer<Float> dst, TransformInfo transformInfo)
-        {
-            ectx.Check(src.Length == transformInfo.Dimension);
+                public int InputIndex
+                {
+                    get
+                    {
+                        // Column names are already checked by PcaTransform
+                        _schema.TryGetColumnIndex(_input, out int index);
+                        return index;
+                    }
+                }
+
+                public ColumnType InputType => _schema[_input].Type;
```

Comment:
>public ColumnType InputType => _schema[_input].Type; [](start = 16, length = 52)

instead of properties over internal state, let's adopt a pattern where all of the properties are computed once in the ctor, and there is no private fields.